src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,564 @@
+
+    Returns:
+        `torch.nn.Module`: The Vera model.
+
+    Example:
+
+        ```py
+        >>> from transformers import AutoModelForSeq2SeqLM
+        >>> from peft import VeraModel, VeraConfig
+
+        >>> config = VeraConfig(
+        ...     task_type="SEQ_2_SEQ_LM",
+        ...     r=8,
+        ...     target_modules=["q", "v"],
+        ...     vera_dropout=0.01,
+        ... )
+
+        >>> model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")
+        >>> vera_model = VeraModel(model, config, "default")
+        ```
+
+        ```py
+        >>> import transformers
+        >>> from peft import VeraConfig, PeftModel, get_peft_model
+
+        >>> target_modules = ["q_proj", "k_proj", "v_proj", "out_proj", "fc_in", "fc_out", "wte"]
+        >>> config = VeraConfig(
+        ...     r=4, target_modules=target_modules, vera_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+        ... )
+
+        >>> model = transformers.GPTJForCausalLM.from_pretrained(
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
+    def _find_first_dim(self) -> int:
+        """
+        Finds the first linear and embedding that has been wrapped with Vera, and extract the input and output
+        dimension.
+
+        This will be used for determining the size of the shared vera_A and vera_B matrices.
+
+        This will throw an error if there are multiple layers of the same type with different shapes.
+        """
+        first_linear, first_embedding = None, None
+        for module in self.model.modules():
+            if isinstance(module, Linear):
+                module_shape = tuple(module.weight.shape)
+                if module.fan_in_fan_out:
+                    module_shape = module_shape[::-1]
+
+                if first_linear is not None and module_shape != first_linear:
+                    raise ValueError(
+                        "Multiple target linear layers with different dimensions were specified! Vera only supports a"
+                        f" single dimension size. Got '{module_shape}' expected '{first_linear}"
+                    )
+                first_linear = module_shape
+
+            elif isinstance(module, Embedding):
+                if first_embedding is not None and tuple(module.weight.shape) != first_embedding:
```

Comment:
Just so I'm sure I get it right, the idea is that we can have a distinct `vera_A` and `vera_B` for each type of layer (currently only `Linear` and `Embedding`), is that right?

An uncovered use case that comes to my mind would be if we want to have multiple vera adapters with different dimensions on the same model, this would currently fail. Even if they have same dimension, I wonder if we shouldn't have distinct `vera_A` and `vera_B` for each adapter, as they could come from different sources that each trained with different seeds. If you agree, I think we need to distinguish by adapter name, and `vera_A` and `vera_B` should be `nn.ModuleDict`s.

Reply:
Your understanding of the current method is correct ~

To be honest, I am not familiar with using multiple adapters / why you would do so. Could you outline what would need to be changed to support this?

Reply:
So first the why: Let's imagine that a user has multiple Vera adapters and wants to run inference with all of them. If we enable multiple adapters, they can load the base mode once, add all Vera adapters, and switch between them between batches. If only a single adapter were allowed, the user would have to load the base model into memory once per adapter. On top, we can also enable activating multiple adapters at once.

Regarding the how: Instead of having `vera_A` and `vera_B` be tensors, they would be `nn.ModuleDict`s that contain 1 to n `vera_A` and `vera_B` each. Each time that currently `self.vera_A` is being fetched, we would need to fetch `self.vera_A[active_adapter]`. Note that a lot of the code already takes multiple active adapters into account, e.g. when doing `for active_adapter in self.active_adapters` in `forward`. 

One ugly problem with this though is that `nn.ModuleDict` cannot be assigned as a buffer. Therefore, this cannot be changed in a straightforward fashion right now :(

I'm not sure what the best solution is. We could maybe make them normal parameters instead of buffers, and then filter them out when the `state_dict` is being fetched. But this may not be enough. Alternatively, we could just not allow multiple adapters to be added, which would be unfortunate. In that case, we should raise an explicit error when trying to load a second adapter. Do you have an idea?

Reply:
Ideally, we would have something [like this](https://github.com/pytorch/pytorch/issues/37386) to store buffers in :sweat_smile: perhaps we can make something ourselves that could achieve the same effect?

Reply:
For example, `botorch` has it https://botorch.org/api/_modules/botorch/utils/torch.html

Reply:
Interesting finding. In principle, I wouldn't be opposed to copying that implementation, as it is (hopefully) battle-tested. Most importantly, if this works right away with saving and loading `state_dict`s, that would be great.

Reply:
I've added this in the most recent commits, take a look! :slightly_smiling_face: